log.go

```diff
@@ -2,12 +2,119 @@ package didplc
 
 import (
 	"fmt"
+	"sync"
 	"time"
 
-	"github.com/bluesky-social/indigo/atproto/crypto"
 	"github.com/bluesky-social/indigo/atproto/syntax"
 )
 
+type OpStatus struct {
+	CreatedAt   time.Time // the only immutable field here
+	Nullified   bool
+	LastChild   string // CID
+	AllowedKeys []string
+}
+
+// Note: LogValidationContext is designed such that it could later be turned into an interface,
+// optionally backed by a db rather than in-memory
+// Note: ops are globally unique by CID, so opStatus map can be shared across all DIDs
+type LogValidationContext struct {
+	head     map[string]string    // DID -> CID, tracks most recent valid op for a particular DID
+	opStatus map[string]*OpStatus // CID -> OpStatus
+	lock     sync.RWMutex
```

Comment:
This lock is currently superfluous since, at present, code exercising `LogValidationContext` only uses it on a single thread.

I mainly added it to help me think about how locking *would* work under a more concurrent usage scenario.